hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -712,6 +717,245 @@ case class PartitionNativeReader(spec: AbstractTypedCodecSpec, uidFieldName: Str
+      val ib = mb.genFieldThisRef[InputBuffer]("buffer")
+
+      // leave the index open/initialized to allow queries to reuse the same index for the same file
+      val indexInitialized = mb.genFieldThisRef[Boolean]("index_init")
+
+      val region = mb.genFieldThisRef[Region]("pnr_region")
+
+      val decodedRow = cb.emb.newPField("rowsValue", eltSType)
+
+      val producer = new StreamProducer {
+        override def method: EmitMethodBuilder[_] = mb
+        override val length: Option[EmitCodeBuilder => Code[Int]] = None
+
+        override def initialize(cb: EmitCodeBuilder, outerRegion: Value[Region]): Unit = {
+
+          val startBound = ctx.loadStart(cb).get(cb)
+          val includesStart = ctx.includesStart()
+          val endBound = ctx.loadEnd(cb).get(cb)
+          val includesEnd = ctx.includesEnd()
+
+          val startAndSignTuple = SStackStruct.constructFromArgs(cb,
+            region,
+            TTuple(startBound.st.virtualType, TInt32),
+            EmitCode.present(cb.emb, startBound),
+            EmitCode.present(cb.emb, primitive(cb.ifx[Int](includesStart, -1, 1)))
+          )
+
+          val endAndSignTuple = SStackStruct.constructFromArgs(cb,
+            region,
+            TTuple(endBound.st.virtualType, TInt32),
+            EmitCode.present(cb.emb, endBound),
+            EmitCode.present(cb.emb, primitive(cb.ifx[Int](includesEnd, 1, -1)))
+          )
+
+          def compareEnd(cb: EmitCodeBuilder, containerEltEV: EmitValue, partBoundEV: EmitValue, f: Value[Int] => Code[Boolean]): Value[Boolean] = {
+            val containerElt = containerEltEV.get(cb).asInterval
+            val partitionBoundStart = containerElt.loadEnd(cb).get(cb).asBaseStruct
+            val partitionBoundSign = cb.ifx[Int](containerElt.includesEnd(), 1, -1)
+            val queryBound = partBoundEV.get(cb).asBaseStruct
+            val queryPoint = queryBound.loadField(cb, 0).get(cb).asBaseStruct
+            val querySign = queryBound.loadField(cb, 1).get(cb).asInt.value
+            cb.memoize(f(partitionIntervalEndpointCompare(cb, partitionBoundStart, partitionBoundSign, queryPoint, querySign)))
+          }
+
+          cb.assign(startPartitionIndex, new BinarySearch(cb.emb,
+            partitionerRuntime.st,
+            EmitType(startAndSignTuple.st, true),
+            ((cb, elt) => elt),
+            bound = "lower",
+            compareEnd(_, _, _, _ < 1)
+          ).search(cb, partitionerRuntime, EmitCode.present(cb.emb, startAndSignTuple)))
+          cb.assign(currPartitionIdx, startPartitionIndex)
+
+          cb.assign(lastIncludedPartitionIdx, new BinarySearch(cb.emb,
+            partitionerRuntime.st,
+            EmitType(endAndSignTuple.st, true),
+            ((cb, elt) => elt),
+            bound = "lower",
+            compareEnd(_, _, _, _ < 1)
+          ).search(cb, partitionerRuntime, EmitCode.present(cb.emb, endAndSignTuple)))
```

Comment:
From a quick look, I think you've reimplemented the registered function `partitionerFindIntervalRange`. If so, you should just factor out the implementation to a staged function (as most of the interval registered functions are) and call that. I think we should try to keep all the staged interval ordering stuff in one place, since it's relatively complicated.